File: android/src/main/kotlin/ai/synheart/session/SynheartSessionPlugin.kt
package ai.synheart.session

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

/**
 * Flutter plugin entry point for Synheart Session on Android.
 *
 * After the refactor to Dart-side `LiveSessionEngine`, this plugin only
 * handles the Wear OS watch relay path. All local session compute is done
 * in Dart. The watch relay is kept because the Wearable Data Layer requires
 * a native host.
 */
class SynheartSessionPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var watchRelay: WatchSessionRelay? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "ai.synheart.session/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "ai.synheart.session/events")
        eventChannel.setStreamHandler(this)

        watchRelay = WatchSessionRelay(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        watchRelay?.dispose()
        watchRelay = null
    }

    // -- MethodCallHandler ----------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startSession" -> handleStartSession(call, result)
            "stopSession" -> handleStopSession(call, result)
            "getStatus" -> result.success(null) // Local status handled in Dart
            "getWatchStatus" -> handleGetWatchStatus(result)
            else -> result.notImplemented()
        }
    }

    private fun handleStartSession(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        if (args == null) {
            result.error("invalid_state", "Invalid arguments", null)
            return
        }

        @Suppress("UNCHECKED_CAST")
        val config = args as Map<String, Any?>

        val relay = watchRelay
        if (relay == null) {
            result.success(null)
            return
        }

        val sink: (Map<String, Any?>) -> Unit = { event ->
            eventSink?.success(event)
        }

        relay.checkReachable { reachable ->
            if (reachable) {
                relay.startSession(config, sink)
            }
            // If watch not reachable, just return — Dart LiveSessionEngine handles locally
            result.success(null)
        }
    }

    private fun handleStopSession(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val sessionId = args?.get("session_id") as? String
        if (sessionId == null) {
            result.error("invalid_state", "Missing session_id", null)
            return
        }

        watchRelay?.stopSession(sessionId)
        result.success(null)
    }

    private fun handleGetWatchStatus(result: MethodChannel.Result) {
        val relay = watchRelay
        if (relay == null) {
            result.success(mapOf(
                "supported" to false,
                "reachable" to false,
                "paired" to false,
                "installed" to false
            ))
            return
        }

        relay.getStatus { status ->
            result.success(status)
        }
    }

    // -- EventChannel.StreamHandler -------------------------------------------

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Classes/SynheartSessionPlugin.swift
import Flutter

/// Flutter plugin entry point for Synheart Session on iOS.
///
/// After the refactor to Dart-side `LiveSessionEngine`, this plugin only
/// handles the Apple Watch relay path. All local session compute is done
/// in Dart. The watch relay is kept because WCSession requires a native host.
public class SynheartSessionPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private let watchRelay = WatchSessionRelay()
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SynheartSessionPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "ai.synheart.session/methods",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "ai.synheart.session/events",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - FlutterPlugin (MethodChannel)

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startSession":  handleStartSession(call: call, result: result)
        case "stopSession":   handleStopSession(call: call, result: result)
        case "getStatus":     result(nil) // Local status handled in Dart
        case "getWatchStatus": handleGetWatchStatus(result: result)
        default:              result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Start (watch relay only)

    private func handleStartSession(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "invalid_state", message: "Invalid arguments", details: nil))
            return
        }

        do {
            let config = try SessionConfig(from: args)
            let sink: ([String: Any]) -> Void = { [weak self] event in
                DispatchQueue.main.async { self?.eventSink?(event) }
            }

            if watchRelay.isReachable {
                watchRelay.startSession(config: config, callback: sink) { /* onSendFailed — Dart handles locally */ }
            }
            // If watch not reachable, just return — Dart LiveSessionEngine handles locally
            result(nil)
        } catch let error as SessionError {
            result(FlutterError(code: error.code.rawValue, message: error.description, details: nil))
        } catch {
            result(FlutterError(code: "invalid_state", message: error.localizedDescription, details: nil))
        }
    }

    // MARK: - Stop

    private func handleStopSession(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sessionId = args["session_id"] as? String else {
            result(FlutterError(code: "invalid_state", message: "Missing session_id", details: nil))
            return
        }

        watchRelay.stopSession(sessionId: sessionId)
        result(nil)
    }

    // MARK: - Watch status

    private func handleGetWatchStatus(result: @escaping FlutterResult) {
        result(watchRelay.status)
    }

    // MARK: - FlutterStreamHandler (EventChannel)

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
